tiledb/sm/query/writers/global_order_writer.cc

```diff
@@ -1210,74 +1331,94 @@ Status GlobalOrderWriter::prepare_full_tiles_var(
-  if (client) {
-    return {Status::Ok(), global_write_state_->multipart_upload_state_};
+uint64_t GlobalOrderWriter::num_tiles_to_write(
+    uint64_t start,
+    uint64_t tile_num,
+    std::unordered_map<std::string, WriterTileVector>& tiles) {
+  // Cache variables to prevent map lookups.
+  const auto buf_names = buffer_names();
+  std::vector<bool> var_size;
+  std::vector<bool> nullable;
+  std::vector<WriterTileVector*> writer_tile_vectors;
+  var_size.reserve(buf_names.size());
+  nullable.reserve(buf_names.size());
+  writer_tile_vectors.reserve(buf_names.size());
+  for (auto& name : buf_names) {
+    var_size.emplace_back(array_schema_.var_size(name));
+    nullable.emplace_back(array_schema_.is_nullable(name));
+    writer_tile_vectors.emplace_back(&tiles[name]);
   }
 
-  auto meta = global_write_state_->frag_meta_;
-  std::unordered_map<std::string, VFS::MultiPartUploadState> result;
-
-  // TODO: to be refactored, there are multiple places in writers where
-  // we iterate over the internal fragment files manually
-  const auto buf_names = buffer_names();
-  for (const auto& name : buf_names) {
-    auto&& [st1, uri] = meta->uri(name);
-    RETURN_NOT_OK_TUPLE(st1, {});
+  // Make sure we don't write more than the desired fragment size.
+  for (uint64_t t = start; t < tile_num; t++) {
+    uint64_t tile_size = 0;
+    for (uint64_t a = 0; a < buf_names.size(); a++) {
+      if (var_size[a]) {
+        tile_size += writer_tile_vectors[a]
+                         ->at(t)
+                         .offset_tile()
+                         .filtered_buffer()
+                         .size();
+        tile_size +=
+            writer_tile_vectors[a]->at(t).var_tile().filtered_buffer().size();
+      } else {
+        tile_size +=
+            writer_tile_vectors[a]->at(t).fixed_tile().filtered_buffer().size();
+      }
 
-    auto&& [st2, state] = storage_manager_->vfs()->multipart_upload_state(*uri);
-    RETURN_NOT_OK_TUPLE(st2, {});
-    // If there is no entry for this uri, probably multipart upload is disabled
-    // or no write was issued so far
-    if (!state.has_value()) {
-      return {Status::Ok(), {}};
+      if (nullable[a]) {
+        tile_size += writer_tile_vectors[a]
+                         ->at(t)
+                         .validity_tile()
+                         .filtered_buffer()
+                         .size();
+      }
     }
-    result[uri->remove_trailing_slash().last_path_part()] = std::move(*state);
 
-    if (array_schema_.var_size(name)) {
-      auto&& [status, var_uri] = meta->var_uri(name);
-      RETURN_NOT_OK_TUPLE(status, {});
-
-      auto&& [st, var_state] =
-          storage_manager_->vfs()->multipart_upload_state(*var_uri);
-      RETURN_NOT_OK_TUPLE(st, {});
-      result[var_uri->remove_trailing_slash().last_path_part()] =
-          std::move(*var_state);
+    if (current_fragment_size_ + tile_size > fragment_size_) {
+      current_fragment_size_ = 0;
```

Comment:
Nit: Don't reset fragment size here. Wait until you've actually created a new fragment with fragments size 0. I suggest creating a function that both creates the new fragment and sets `current_fragment_size_ = 0`.

Reply:
Moved setting the variable next to the create_fragment call.